### PR TITLE
Expand kubelet's clientset to support multiple tenant partitions

### PR DESCRIPTION
**Pull request by yb01**
Initial code change to expand kubelet to support multiple Tenant partitions and separation of Resource and Tenant partition clientsets.

make and run:

run it with new args:
/kubelet --kubeconfig=/testconfig/kubeconfig1 --tenant-servers="http://10.1.1.1:8080,http://10.1.1.2:8080" > kubelet.log 2>&1 &
OR to set up the resource cluster with:
export TENANT_SERVERS="http://172.31.10.115:8080,http://172.31.14.52:8080" && export SCALE_OUT_PROXY_ENDPOINT="http://172.31.10.115:8888" && LOG_LEVEL=3 KUBE_CONTROLLERS=nodelifecycle IS_RESOURCE_PARTITION=true ./hack/arktos-up-scale-out-poc.sh
 
Noted that it uses a new cmdline args for now. add new entries to kubecofig will involved in the client-go config update which currently i incline to not change.

```
Below is a quick testing with them. Issue #1 and #2 are debatable if we by-design or treat them as bugs. I am looking into issue 3.
 
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig1 get nodes
No resources found.
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig2 get nodes
No resources found.
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=proxkubeconfig get nodes
error: stat proxkubeconfig: no such file or directory
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=proxykubeconfig get nodes
NAME              STATUS   ROLES    AGE     VERSION
ip-172-31-8-177   Ready    <none>   5m47s   v0.0.0-master+$Format:%h$
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=resourcekubeconfig get nodes
NAME              STATUS   ROLES    AGE     VERSION
ip-172-31-8-177   Ready    <none>   5m59s   v0.0.0-master+$Format:%h$
root@ip-172-31-10-115:/testconfig# 
 
issue 1: tenant creation should be directed to desired partition
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=proxykubeconfig create -f /yamls/tenant1.yaml 
tenant/a created
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=proxykubeconfig create -f /yamls/tenant2.yaml 
tenant/x created
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=proxykubeconfig get tenants
NAME     STORAGEID   STATUS   AGE
a        1           Active   20s
system   0           Active   8m54s
x        1           Active   12s
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig1 get tenants
NAME     STORAGEID   STATUS   AGE
a        1           Active   34s
system   0           Active   9m8s
x        1           Active   26s
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig2 get tenants
NAME     STORAGEID   STATUS   AGE
system   0           Active   6m31s
root@ip-172-31-10-115:/testconfig# 
 
issue 2: get tenant from proxy should return all, currently return the first partition’s tenant
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig2 create -f /yamls/tenant2.yaml 
tenant/x created
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig1 create -f /yamls/tenant1.yaml 
tenant/a created
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=proxykubeconfig get tenants
NAME     STORAGEID   STATUS   AGE
a        1           Active   4s
system   0           Active   11m
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig1 get tenants
NAME     STORAGEID   STATUS   AGE
a        1           Active   13s
system   0           Active   11m
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig2 get tenants
NAME     STORAGEID   STATUS   AGE
system   0           Active   8m48s
x        1           Active   25s
root@ip-172-31-10-115:/testconfig# 
 
issue 3: pods from second partition remains in scheduled and pending:
root@ip-172-31-10-115:/testconfig# vi /yamls/condefault.yaml 
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig2 create -f /yamls/condefault2.yaml 
pod/condefault2 created
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig1 create -f /yamls/condefault1.yaml 
pod/condefault created
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig1 get pods
No resources found.
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig1 get pods -AT
TENANT   NAMESPACE     NAME                        HASHKEY               READY   STATUS             RESTARTS   AGE
a        default       condefault                  3123577652425193689   1/1     Running            0          11s
system   kube-system   kube-dns-6fdcc959b4-wz5p6   1151037929471017490   0/3     CrashLoopBackOff   0          15m
system   kube-system   virtlet-568zj               5227541762739606200   0/3     CrashLoopBackOff   0          14m
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig2 get pods -AT
TENANT   NAMESPACE     NAME                        HASHKEY               READY   STATUS    RESTARTS   AGE
system   kube-system   kube-dns-6fdcc959b4-htp2h   1585220668317444351   0/3     Pending   0          12m
system   kube-system   virtlet-kc9dg               6279963007203658651   0/3     Pending   0          12m
x        default       condefault2                 2135629449184201393   0/1     Pending   0          30s
 
 
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2020-11-12T22:06:58Z"
      status: "True"
      type: PodScheduled
    phase: Pending
    qosClass: Guaranteed
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
root@ip-172-31-10-115:/testconfig# 

 
235767 I1112 22:45:25.109715   28480 kubelet_pods.go:1336] Generating status for "condefault3_default_x(e7a816c5-4094-4f6a-95ff-99d6dcd8f220)"
235768 I1112 22:45:25.109767   28480 kubelet_pods.go:1435] pod not found in the status manager map
235769 I1112 22:45:25.112077   28480 status_manager.go:484] Pod "condefault3" (e7a816c5-4094-4f6a-95ff-99d6dcd8f220) does not exist on the server
235770 I1112 22:45:25.129183   28480 factory.go:173] Using factory "raw" for container "/kubepods/pode7a816c5-4094-4f6a-95ff-99d6dcd8f220"
 
Issue #4: pod from first partition, which should have the status etc as we default to first partition, crashed after running a while.
root@ip-172-31-10-115:/testconfig# kubectl --kubeconfig=kubeconfig1 get pods -AT
TENANT   NAMESPACE     NAME                        HASHKEY               READY   STATUS             RESTARTS   AGE
a        default       condefault                  3123577652425193689   0/1     CrashLoopBackOff   4          45m
system   kube-system   kube-dns-6fdcc959b4-wz5p6   1151037929471017490   1/3     CrashLoopBackOff   15         60m
system   kube-system   virtlet-568zj               5227541762739606200   0/3     CrashLoopBackOff   18         59m
root@ip-172-31-10-115:/testconfig#

Issue 5. Event is tied with tenant:
10757 E1113 17:19:34.716840   10490 event.go:241] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"condefault.164720e8640b21c9", GenerateName:"", Tenant:"a", Namespace:"default", SelfLink:""      , UID:"", HashKey:0, ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(n      il), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kin      d:"Pod", Namespace:"default", Name:"condefault", UID:"cb7b71f9-b86a-4157-93a0-e67e80f7a554", APIVersion:"v1", ResourceVersion:"841633221399019521", FieldPath:"spec.containers{nginx}", Tenant:"a"}, Reason:"Pulled", Message:"Container image \"ngin      x\" already present on machine", Source:v1.EventSource{Component:"kubelet", Host:"ip-172-31-8-177"}, FirstTimestamp:v1.Time{Time:time.Time{wall:0xbfe3cde9a79125c9, ext:270625226321, loc:(*time.Location)(0xc99ba20)}}, LastTimestamp:v1.Time{Time:t      ime.Time{wall:0xbfe3cde9a79125c9, ext:270625226321, loc:(*time.Location)(0xc99ba20)}}, Count:1, Type:"Normal", EventTime:v1.MicroTime{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Series:(*v1.EventSeries)(nil), Action:"", Related:      (*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'tenants "a" not found' (will not retry!)

```

### CODE REVIEW

File: hack/lib/common.sh [hunk 1]
```diff
@@ -457,7 +457,7 @@ function kube::common::start_workload_controller_manager {
 function kube::common::start_controller_manager {
     local controller_opts="*,-nodelifecycle,-nodeipam"
     if [ "${IS_RESOURCE_PARTITION}" == "true" ]; then
-       controller_opts="nodelifecycle,nodeipam"
+       controller_opts="nodelifecycle"
```
Comment (pdgetrf) [inline hack/lib/common.sh]: looks like your code is not based on the most recent code

Comment (yb01) [reply hack/lib/common.sh]: yeah, this is merged and removed.

**Comment by pdgetrf** [comment]
I took this and run with just one apiserver setup like this:

image

And pod is schedule but not running:

```java
root@goose-t:~/code/src/k8s.io/arktos# kubectl get pods -o wide
NAME                     HASHKEY               READY   STATUS    RESTARTS   AGE   IP       NODE      NOMINATED NODE   READINESS GATES
nginx-7fc6c5979b-dczfq   5557352906362607514   0/1     Pending   0          68s   <none>   goose-r   <none>           <none>
```

In kublet log I'm seeing
```
E1113 01:14:13.006435   19696 reflector.go:407] k8s.io/kubernetes/pkg/kubelet/kubelet.go:454: Failed to watch *v1.Node: Get http://192.168.0.120:8888/api/v1/nodes?fieldSelector=metadata.name%3Dgoose-r&resourceVersion=841602838005547008&timeout=10s&timeoutSeconds=557&watch=true: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

I think we want to fix this before merging.

**Review by pdgetrf — APPROVED** [review]
verified with single tenant apiserver in a single-tenant-partition env. pod can stay running without problem.

**Comment by pdgetrf** [comment]
@yb01 can you also make the script change in this? say taking the tenant partition ip addresses from an environment. otherwise it may cause regression.

**Review by pdgetrf — CHANGES_REQUESTED** [review]
please make the arktos-up script change too

**Comment by yb01** [comment]
> please make the arktos-up script change too
just pushed a change to the arktos-up-scale-out.sh to set the defualt for the --tenant-services parameters, to not break the existing cluster set up.

**Review by pdgetrf — APPROVED** [review]
looks good to me.

please add a "validation" section in the description with info on how this is tested. mainly the following:

1. any known regression (breaking change?)
2. what is the scenario that's been tested
3. any other known issue (set up, etc)

**Comment by yb01** [comment]
> looks good to me.
> 
> please add a "validation" section in the description with info on how this is tested. mainly the following:
> 
> 1. any known regression (breaking change?)
> 2. what is the scenario that's been tested
> 3. any other known issue (set up, etc)

Update the PR with test info and results with issues.

**Pull request merged**